src/util/tev_curl.c

```diff
@@ -0,0 +1,896 @@
+
+        if (socketp == NULL) {
+            /* If this socket doesn't have private data, it must be a new one,
+             * let's start tracking it with tevent
+             */
+            tcsock = register_curl_socket(tctx, s, flags);
+            if (tcsock == NULL) {
+                return 1;
+            }
+        } else {
+            /* If we are already tracking this socket, just set the correct
+             * flags for tevent and pass the control to tevent
+             */
+            tcsock = talloc_get_type(socketp, struct tcurl_sock);
+            if (tcsock == NULL) {
+                DEBUG(SSSDBG_CRIT_FAILURE,
+                      "BUG: No private data for socket %d\n", s);
+                return 1;
+            }
+            tevent_fd_set_flags(tcsock->fde, flags);
+        }
+        break;
+
+    case CURL_POLL_REMOVE:
+        /* This socket is being closed by curl, so we need to.. */
+        tcsock = talloc_get_type(socketp, struct tcurl_sock);
+        if (tcsock == NULL) {
+            DEBUG(SSSDBG_CRIT_FAILURE,
+                  "BUG: Trying to remove an untracked socket %d\n", s);
+        }
+        /* ..stop tracking the socket with the multi handle.. */
+        curl_multi_assign(tctx->multi_handle, s, NULL);
+        /* ..and stop tracking the fd with tevent */
+        talloc_free(tcsock);
+        break;
+
+    default:
+        return 1;
+  }
+
+  return 0;
+}
+
+static void check_fd_activity(struct tevent_context *ev,
+                              struct tevent_timer *te,
+                              struct timeval current_time,
+                              void *private_data)
+{
+    struct tcurl_ctx *tctx = talloc_get_type(private_data, struct tcurl_ctx);
+    int running_handles;
+
+    curl_multi_socket_action(tctx->multi_handle,
+                             CURL_SOCKET_TIMEOUT,
+                             0,
+                             &running_handles);
+    DEBUG(SSSDBG_TRACE_ALL,
+          "Still tracking %d outstanding requests\n", running_handles);
+
+    /* https://ec.haxx.se/libcurl-drive-multi-socket.html */
+    process_curl_activity(tctx);
+}
+
+static int schedule_fd_processing(CURLM *multi,
+                                  long timeout_ms,
+                                  void *userp)
+{
+    struct timeval tv = { 0, 0 };
+    struct tcurl_ctx *tctx = talloc_get_type(userp, struct tcurl_ctx);
+
+    if (timeout_ms == -1) {
+        /* man curlmopt_timerfunction(3) says:
+         *  A timeout_ms value of -1 means you should delete your timer.
+         */
+        talloc_zfree(tctx->initial_timer);
+    }
+
+    tv = tevent_timeval_current_ofs(0, timeout_ms * 10);
+
+    tctx->initial_timer = tevent_add_timer(tctx->ev, tctx, tv,
+                                           check_fd_activity, tctx);
+    if (tctx->initial_timer == NULL) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static int tcurl_ctx_destroy(TALLOC_CTX *ptr)
+{
```

Comment:
You can use: `static inc tcurl_ctx_destroy(struct tcurl_ctx *ctx)` directly since talloc makes it type safe during compilation. Then you need to avoid explicit cast to `TALLOC_CTX*` when setting it. The same for `tcurl_http_cleanup_handle()`.